cephfs/snap_diff.go

```diff
@@ -0,0 +1,251 @@
+)
+
+// SnapDiffInfo is a handle to a snapshot diff API.
+type SnapDiffInfo struct {
+	cMount *MountInfo
+	dir1   *Directory
+	dirAux *Directory
+}
+
+// SnapDiffEntry is a single entry in the snapshot diff.
+// It contains a DirEntry and the ID of the snapshot to
+// which the directory belongs.
+type SnapDiffEntry struct {
+	DirEntry *DirEntry
+	SnapID   uint64
+}
+
+// SnapDiffConfig is used to define the parameters of a open_snapdiff call.
+// Snapshot Delta is generated between the passed snapshots snap1 and snap2.
+// All fields must be specified before passing to OpenSnapDiff().
+type SnapDiffConfig struct {
+	// CMount is the ceph mount handle that will be used for snap.diff retrieval.
+	CMount *MountInfo
+	// RootPath represents the root path for snapshots-in-question.
+	RootPath string
+	// RelPath is the subpath under the root to build delta for.
+	RelPath string
+	// Snap1 is the first snapshot name.
+	Snap1 string
+	// Snap2 is the second snapshot name.
+	Snap2 string
+}
+
+// OpenSnapDiff opens a snapshot diff stream to get snapshots delta
+// and returns a SnapDiffInfo struct containing the diff information.
+//
+// Implements:
+//
+//	int ceph_open_snapdiff(struct ceph_mount_info* cmount,
+//	                       const char* root_path,
+//	                       const char* rel_path,
+//	                       const char* snap1,
+//	                       const char* snap2,
+//	                       struct ceph_snapdiff_info* out);
+func OpenSnapDiff(config SnapDiffConfig) (*SnapDiffInfo, error) {
+	if config.CMount == nil || config.RootPath == "" || config.RelPath == "" ||
+		config.Snap1 == "" || config.Snap2 == "" {
+		return nil, fmt.Errorf("invalid arguments")
+	}
+
+	cephOpenSnapDiffOnce.Do(func() {
+		cephOpenSnapDiff, cephOpenSnapDiffErr = dlsym.LookupSymbol("ceph_open_snapdiff")
+	})
+
+	if cephOpenSnapDiffErr != nil {
+		return nil, fmt.Errorf("%w: %w", ErrNotImplemented, cephOpenSnapDiffErr)
+	}
+
+	rawCephSnapDiffInfo := &C._ceph_snapdiff_info{}
+
+	ret := C.open_snapdiff_dlsym(
+		cephOpenSnapDiff,
+		config.CMount.mount,
+		C.CString(config.RootPath),
+		C.CString(config.RelPath),
+		C.CString(config.Snap1),
+		C.CString(config.Snap2),
+		rawCephSnapDiffInfo)
+
+	if ret != 0 {
+		return nil, getError(ret)
+	}
+
+	mountInfo := &MountInfo{
+		mount: rawCephSnapDiffInfo.cmount,
+	}
+	cephSnapDiffInfo := &SnapDiffInfo{
+		cMount: mountInfo,
+		dir1: &Directory{
+			mount: mountInfo,
+			dir:   rawCephSnapDiffInfo.dir1,
+		},
+		dirAux: &Directory{
+			mount: mountInfo,
+			dir:   rawCephSnapDiffInfo.dir_aux,
+		},
+	}
+
+	return cephSnapDiffInfo, nil
+}
+
+// Readdir returns the next entry in the snapshot diff stream.
+// If there are no more entries, it returns (nil, nil).
+//
+// Implements:
+//
+//	int ceph_readdir_snapdiff(struct ceph_snapdiff_info* snapdiff,
+//	                           struct ceph_snapdiff_entry_t* out);
+func (info *SnapDiffInfo) Readdir() (*SnapDiffEntry, error) {
```

Comment:
think about including some validation (other structs have a `.validate()` function), the following should not panic:
```go
var sdi *SnapDiffInfo
_, err := sdi.Readdir()
```